### DISCUSSION TITLE
Some build problems about cheriot llvm.

### DISCUSSION THREAD

**Discussion by waruqi** [CHERIoT-Platform/cheriot-rtos#2] [Q&A] [answered]
I built llvm using cheribuild but it still doesn't seem to support cheriot.

```console
/Users/ruki/projects/working/rtos/cheri/output/sdk/bin/clang -c -Qunused-arguments -target riscv32-unknown-unknown -mcpu=cheriot -mabi=cheriot -mxcheri-rvc -mrelax -fshort-wchar -nostdinc -Oz -g -fomit-frame-pointer -fno-builtin -fno-exceptions -fno-asynchronous-unwind-tables -fno-rtti -Werror -I/Users/ruki/projects/working/rtos/cheriot-rtos/sdk/include/c++-config -I/Users/ruki/projects/working/rtos/cheriot-rtos/sdk/include/libc++ -I/Users/ruki/projects/working/rtos/cheriot-rtos/sdk/include -DCHERIOT_AVOID_CAPRELOCS -DDEBUG_LOADER=false -DSAIL -DCPU_TIMER_HZ=2000 -DTICK_RATE_HZ=10 -DTEMPORAL_SAFETY -DSOFTWARE_REVOKER -DSIMULATION -DDEVICE_EXISTS_clint -DDEVICE_EXISTS_uart -DDEVICE_EXISTS_shadow -DCONFIG_THREADS={{1,2,1024,5},{2,1,1024,5},{3,1,1024,5},} -DCONFIG_THREADS_ENTRYPOINTS={LA_ABS(__export_test_runner__Z9run_testsv),LA_ABS(__export_thread_pool__Z15thread_pool_runv),LA_ABS(__export_thread_pool__Z15thread_pool_runv),} -DCONFIG_THREADS_NUM=3 -o build/.objs/test-suite.loader/cheriot/cheriot/release/__/sdk/core/loader/boot.S.o ../sdk/core/loader/boot.S
checking for flags (-MMD -MF) ... ok
checking for flags (-fdiagnostics-color=always) ... ok
checking for flags (-fdiagnostics-color=always) ... ok
error: clang-13: error: the clang compiler does not support '-mcpu=cheriot'
```

**Comment by davidchisnall** [comment, marked as answer]
Thanks for trying the project!  

That is not the correct LLVM, you need to either:

 - Build it from the cheriot branch of CTSRD-CHERI/llvm-project
 - Use the devcontainer that we provide.

If you are not using VS Code / GitHub Code Spaces, or another editor that supports dev containers, then you can fetch the container image with `docker pull cheriot.azurecr.io/cheriot/devcontainer`.

**Reply by waruqi** [reply]
> The recommended way to get started with LLVM is by using [cheribuild](https://github.com/CTSRD-CHERI/cheribuild), which will build a working toolchain with a single cheribuild.py llvm.

I previously used `cheribuild.py llvm` to build llvm because I saw that it was the recommended way.

> Build it from the cheriot branch of CTSRD-CHERI/llvm-project

I will try it again, thanks.
